mqtt/mqttd/src/broker/mod.rs

```diff
@@ -52,55 +98,15 @@ where
     info!("persisting state before exiting...");
     persistor.store(state).await?;
     info!("state persisted.");
-    info!("exiting... goodbye");
 
+    info!("exiting... goodbye");
     Ok(())
 }
 
-async fn start_snapshotter(
-    broker_handle: BrokerHandle,
-    persistor: FilePersistor<VersionedFileFormat>,
-    snapshot_interval: Duration,
-) -> (
-    ShutdownHandle,
-    JoinHandle<FilePersistor<VersionedFileFormat>>,
-) {
-    let snapshotter = Snapshotter::new(persistor);
-    let snapshot_handle = snapshotter.snapshot_handle();
-    let shutdown_handle = snapshotter.shutdown_handle();
-    let join_handle = tokio::spawn(snapshotter.run());
-
-    // Tick the snapshotter
-    let tick = tick_snapshot(
-        snapshot_interval,
-        broker_handle.clone(),
-        snapshot_handle.clone(),
-    );
-    tokio::spawn(tick);
-
-    // Signal the snapshotter
-    let snapshot = snapshot::snapshot(broker_handle, snapshot_handle);
-    tokio::spawn(snapshot);
-
-    (shutdown_handle, join_handle)
-}
-
-async fn tick_snapshot(
-    period: Duration,
-    mut broker_handle: BrokerHandle,
-    snapshot_handle: StateSnapshotHandle,
-) {
-    info!("Persisting state every {:?}", period);
-    let start = Instant::now() + period;
-    let mut interval = tokio::time::interval_at(start, period);
-    loop {
-        interval.tick().await;
-        if let Err(e) = broker_handle.send(Message::System(SystemEvent::StateSnapshot(
-            snapshot_handle.clone(),
-        ))) {
-            warn!(message = "failed to tick the snapshotter", error=%e);
-        }
-    }
+fn collapse_snapshot_result(
+    server_output: Result<Result<BrokerSnapshot>, JoinError>,
+) -> Result<BrokerSnapshot> {
+    server_output?
```

Comment:
Any recommendations for how to do this more cleanly? I wanted to use a Result function, but couldn't find a good approach.

Reply:
lets discuss offline?